File: python/side_channels/valid_design_side_channel.py
from mlagents_envs.side_channel.side_channel import (
    SideChannel,
    IncomingMessage,
    OutgoingMessage,
)

import uuid
from custom_await import wait_until

class ValidateDesignSideChannel(SideChannel):
    result = False
    received = False

    def __init__(self) -> None:
        super().__init__(uuid.UUID("621f0a70-4f87-11ea-a6bf-784f4387d1f7"))

    def on_message_received(self, msg: IncomingMessage) -> None:
        self.result = msg.read_bool()
        self.received = True

    def reset(self):
        self.result = False
        self.received = False

    def get_result_wait(self):
        if not wait_until(lambda: self.received, 10):
            return False

        return self.result

    def send_string(self, data: str) -> None:
        # Add the string to an OutgoingMessage
        msg = OutgoingMessage()
        msg.write_string(data)
        # We call this method to queue the data we want to send
        super().queue_message_to_send(msg)

File: UnityProject/Assets/DronesAssembly/Scripts/Channels/ValidDesignSideChannel.cs
using System;
using Unity.MLAgents.SideChannels;
using UnityEngine;

public class ValidDesignSideChannel : SideChannel
{
    public ValidDesignSideChannel()
    {
        ChannelId = new Guid("621f0a70-4f87-11ea-a6bf-784f4387d1f7");
    }

    protected override void OnMessageReceived(IncomingMessage msg)
    { }

    public void SendMessage(string stringToSend)
    {
        using (var msgOut = new OutgoingMessage())
        {
            msgOut.WriteString(stringToSend);
            QueueMessageToSend(msgOut);
        }
    }
    
    public void SendMessage(int value)
    {
        using (var msgOut = new OutgoingMessage())
        {
            msgOut.WriteInt32(value);
            QueueMessageToSend(msgOut);
        }
    }
    
    public void SendMessage(bool value)
    {
        using (var msgOut = new OutgoingMessage())
        {
            msgOut.WriteBoolean(value);
            QueueMessageToSend(msgOut);
        }
    }
}